storage/Devices/MdImpl.cc

```diff
@@ -291,25 +313,136 @@ namespace storage
     }
 
 
+    void
+    Md::Impl::calculate_region_and_topology()
+    {
+	vector<BlkDevice*> devices = get_devices();
+
+	long real_chunk_size = chunk_size;
+
+	if (real_chunk_size == 0)
+	    real_chunk_size = get_default_chunk_size();
+
+	// mdadm uses a chunk size of 64 KiB just in case the RAID1 is ever reshaped to RAID5.
+	if (md_level == RAID1)
+	    real_chunk_size = 64 * KiB;
+
+	int number = 0;
+	unsigned long long sum = 0;
+	unsigned long long smallest = std::numeric_limits<unsigned long long>::max();
+
+	for (const BlkDevice* device : devices)
+	{
+	    // TODO handle spare
+
+	    unsigned long long size = device->get_size();
+
+	    // metadata for version 1.0 is 4 KiB block at end aligned to 4 KiB,
+	    // https://raid.wiki.kernel.org/index.php/RAID_superblock_formats
+	    size = (size & ~(0x1000 - 1)) - 0x2000;
```

Comment:
Is this safe? 0x1000 is an integer constant, yet the calculations are based on the assumption that everything is unsigned long long. IMHO it would be a lot safer to user 0x1000ull here to ensure that type in the innermost statement.